yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,393 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+		_, err = hash.Write([]byte("."))
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	ha(request.Caller)
+	ha(request.Service)
+	ha(string(request.Encoding))
+	ha(request.Procedure)
+
+	headersMap := request.Headers.Items()
+	orderedHeadersKeys := make([]string, 0, len(headersMap))
+	for k := range headersMap {
+		orderedHeadersKeys = append(orderedHeadersKeys, k)
+	}
+	sort.Strings(orderedHeadersKeys)
+	for _, k := range orderedHeadersKeys {
+		ha(k)
+		ha(headersMap[k])
+	}
+
+	ha(request.ShardKey)
+	ha(request.RoutingKey)
+	ha(request.RoutingDelegate)
+
+	_, err := hash.Write(body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return fmt.Sprintf("%x", hash.Sum64())
+}
+
+func (r *Recorder) makeFilePath(request *transport.Request, hash string) string {
+	s := fmt.Sprintf("%s.%s.%s.yaml", request.Service, request.Procedure, hash)
+	return filepath.Join(r.recordsDir, sanitizeFilename(s))
+}
+
+// Call implements the yarpc transport filter interface
+func (r *Recorder) Call(
+	ctx context.Context,
+	request *transport.Request,
+	out transport.Outbound,
+) (*transport.Response, error) {
+	log := r.logger
+	requestBody, err := ioutil.ReadAll(request.Body)
+	request.Body = bytes.NewReader(requestBody)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestHash := r.hashRequest(request, requestBody)
+	filepath := r.makeFilePath(request, requestHash)
+
+	switch r.mode {
+	case Replay:
+		response, err := r.loadRecord(filepath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		return response, nil
+	case Append:
+		response, err := r.loadRecord(filepath)
+		if err == nil {
+			return response, nil
+		}
+		fallthrough
+	case Overwrite:
+		response, err := out.Call(ctx, request)
+		if err == nil {
+			r.saveRecord(request, requestBody, response, filepath)
+		}
+		return response, err
+	default:
+		panic("Invalid mode")
+	}
+}
+
+// loadRecord attempts to load a record from the given file. If the record
+// cannot be found the errRecordNotFound is returned. Any other error will
+// abort the current test.
+func (r *Recorder) loadRecord(filepath string) (*transport.Response, error) {
+	rawRecord, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errRecordNotFound
+		}
+		r.logger.Fatal(err)
+	}
+	var cachedRecord record
+	if err := yaml.Unmarshal(rawRecord, &cachedRecord); err != nil {
+		r.logger.Fatal(err)
+	}
+
+	if cachedRecord.Version != currentRecordVersion {
+		r.logger.Fatal(fmt.Sprintf("unsupported record version %d (expected %d)",
+			cachedRecord.Version, currentRecordVersion))
+	}
+
+	response := transport.Response{
+		Headers: transport.HeadersFromMap(cachedRecord.Response.Headers),
+		Body:    ioutil.NopCloser(bytes.NewReader(cachedRecord.Response.Body)),
+	}
+	return &response, nil
+}
+
+// saveRecord attempts to save a record to the given file, any error fails the
+// current test.
+func (r *Recorder) saveRecord(request *transport.Request, requestBody []byte,
```

Comment:
You don't do anything more with the request and requestBody. I would construct the `requestRecord` in `Call` and pass that in here along with the response.

Better yet it seems like you should separate these out:

```
func requestToRecord(transport.Request) (requestRecord, error) { .. }
func responseToRecord(transport.Response) (responseRecord, error) { .. }
func recordToResponse(responseRecord) transport.Response { .. }
```

Then you can do something like,

```
func (r *Recorder) loadRecord(path string) (requestRecord, responseRecord, error) { .. }
func (r *Recorder) saveRecord(path string, req requestRecord, res responseRecord) error { .. }
```